src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Map.cs

```diff
@@ -60,5 +71,143 @@ public void ReadEndMap()
+
+            if (_isConformanceLevelCheckEnabled && CborConformanceLevelHelpers.RequiresSortedKeys(ConformanceLevel))
+            {
+                ValidateSortedKeyEncoding(currentKeyRange);
+            }
+            else if (_isConformanceLevelCheckEnabled && CborConformanceLevelHelpers.RequiresUniqueKeys(ConformanceLevel))
+            {
+                ValidateKeyUniqueness(currentKeyRange);
+            }
+
+            _curentItemIsKey = false;
+        }
+
+        private void HandleMapValueAdded()
+        {
+            Debug.Assert(_currentKeyOffset != null && !_curentItemIsKey);
+
+            _currentKeyOffset = _bytesRead;
+            _curentItemIsKey = true;
+        }
+
+        private void ValidateSortedKeyEncoding((int Offset, int Length) currentKeyRange)
+        {
+            Debug.Assert(_currentKeyOffset != null);
+
+            if (_previousKeyRange != null)
+            {
+                (int Offset, int Length) previousKeyRange = _previousKeyRange.Value;
+
+                ReadOnlySpan<byte> originalBuffer = _originalBuffer.Span;
+                ReadOnlySpan<byte> previousKeyEncoding = originalBuffer.Slice(previousKeyRange.Offset, previousKeyRange.Length);
+                ReadOnlySpan<byte> currentKeyEncoding = originalBuffer.Slice(currentKeyRange.Offset, currentKeyRange.Length);
+
+                int cmp = CborConformanceLevelHelpers.CompareKeyEncodings(previousKeyEncoding, currentKeyEncoding, ConformanceLevel);
+                if (cmp > 0)
+                {
+                    ResetBuffer(currentKeyRange.Offset);
+                    throw new FormatException("CBOR map keys are not in sorted encoding order.");
+                }
+                else if (cmp == 0 && CborConformanceLevelHelpers.RequiresUniqueKeys(ConformanceLevel))
+                {
+                    ResetBuffer(currentKeyRange.Offset);
+                    throw new FormatException("CBOR map contains duplicate keys.");
+                }
+            }
+
+            _previousKeyRange = currentKeyRange;
+        }
+
+        private void ValidateKeyUniqueness((int Offset, int Length) currentKeyRange)
+        {
+            Debug.Assert(_currentKeyOffset != null);
+
+            HashSet<(int Offset, int Length)> previousKeys = GetPreviousKeyRanges();
+
+            if (!previousKeys.Add(currentKeyRange))
+            {
+                ResetBuffer(currentKeyRange.Offset);
+                throw new FormatException("CBOR map contains duplicate keys.");
+            }
+        }
+
+        private HashSet<(int Offset, int Length)> GetPreviousKeyRanges()
+        {
+            return _previousKeyRanges ??= RequestKeyEncodingRangeSet();
+
+            HashSet<(int Offset, int Length)> RequestKeyEncodingRangeSet()
+            {
+                if (_pooledKeyEncodingRangeSets != null)
+                {
+                    HashSet<(int Offset, int Length)>? result;
+
+                    lock (_pooledKeyEncodingRangeSets)
```

Comment:
I don't think we're trying to make the types be thread-safe; so the only locking/interlocked work I'd expect to see would be around statics (like working with the shared array pool).

Reply:
Did this in the similar vein as other pooled objects in the same class (e.g. the [`_indefiniteLengthStringRangeAllocation` field in CborReader.String.cs](https://github.com/dotnet/runtime/blob/683508940d68cd4d8ed7b2e3eceeffd05134660b/src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.cs#L343)), namely to avoid having the same pooled instance being used in multiple contexts in case of otherwise unsupported concurrent operations. I generally think this is overcompensating though.

Reply:
The heuristic is "using a non-thread-safe instance concurrently is allowed to corrupt that instance, but nothing else" (there's a little bit of fuzziness there, e.g. abusing MemoryStream over a provided buffer is going to corrupt the underlying buffer, but that's OK, you told it to).

If there's concurrent access to this value, what's the worse that happens?

* Inconsistent data? Don't do concurrent operations.
* Weird exceptions? Don't do concurrent operations.
* Double-return to an array pool, causing two future instances to have accidental concurrency problems? We should guard against that, since it was shared state corruption.
* Double-rent from an array pool, leading to the pooled array being garbage collected? Don't do concurrent operations.

Reply:
In that case, I think it would make sense to remove both.

Reply:
Addressed.